File: legacy/services/auth-triggers/src/custom-message.ts
import type { CustomMessageTriggerEvent } from 'aws-lambda';

export async function handler(event: CustomMessageTriggerEvent): Promise<CustomMessageTriggerEvent> {
  try {
    const triggerSource = event.triggerSource;
    const code = event.request.codeParameter;
    const name = event.request.userAttributes.name ?? 'ka-TowCommand';

    if (triggerSource === 'CustomMessage_SignUp') {
      event.response.smsMessage =
        `[TowCommand PH] Mabuhay, ${name}! Your verification code is: ${code}. ` +
        `Valid for 15 minutes. Do not share this code.`;
      event.response.emailSubject = 'Welcome to TowCommand PH - Verify Your Account';
      event.response.emailMessage =
        `<h2>Mabuhay, ${name}!</h2>` +
        `<p>Welcome to TowCommand PH - Ang Grab ng Towing!</p>` +
        `<p>Your verification code is: <strong>${code}</strong></p>` +
        `<p>This code expires in 15 minutes.</p>` +
        `<p>If you did not create an account, please ignore this email.</p>` +
        `<p>- Team TowCommand</p>`;
    }

    if (triggerSource === 'CustomMessage_ForgotPassword') {
      event.response.smsMessage =
        `[TowCommand PH] Your password reset code is: ${code}. ` +
        `Valid for 15 minutes. If you did not request this, contact support.`;
      event.response.emailSubject = 'TowCommand PH - Password Reset';
      event.response.emailMessage =
        `<h2>Password Reset Request</h2>` +
        `<p>Hi ${name}, we received a request to reset your password.</p>` +
        `<p>Your reset code is: <strong>${code}</strong></p>` +
        `<p>This code expires in 15 minutes.</p>` +
        `<p>If you did not request this, please ignore this email and your password will remain unchanged.</p>`;
    }

    if (triggerSource === 'CustomMessage_ResendCode') {
      event.response.smsMessage =
        `[TowCommand PH] Your new verification code is: ${code}. Valid for 15 minutes.`;
      event.response.emailSubject = 'TowCommand PH - New Verification Code';
      event.response.emailMessage =
        `<p>Hi ${name}, here is your new verification code: <strong>${code}</strong></p>` +
        `<p>This code expires in 15 minutes.</p>`;
    }

    if (triggerSource === 'CustomMessage_Authentication') {
      event.response.smsMessage =
        `[TowCommand PH] Your login code is: ${code}. Do not share this with anyone.`;
    }

    return event;
  } catch (error) {
    console.error('Custom message error:', error);
    throw error;
  }
}

File: internal/usecase/auth/custom_message.go
package auth

import (
	"context"
	"log/slog"

	"github.com/aws/aws-lambda-go/events"

	"github.com/davidlramirez95/towcommand/internal/adapter/handler"
)

// CustomMessageUseCase handles the Cognito CustomMessage trigger.
// It provides Filipino/English message templates for verification codes,
// password resets, and code resends.
type CustomMessageUseCase struct{}

// NewCustomMessageUseCase creates a CustomMessageUseCase.
func NewCustomMessageUseCase() *CustomMessageUseCase {
	return &CustomMessageUseCase{}
}

// Execute customizes the email and SMS messages for the given trigger source.
func (uc *CustomMessageUseCase) Execute(ctx context.Context, event *events.CognitoEventUserPoolsCustomMessage) (events.CognitoEventUserPoolsCustomMessage, error) {
	triggerSource := handler.CognitoTriggerSource(&event.CognitoEventUserPoolsHeader)
	userName := handler.CognitoUserName(&event.CognitoEventUserPoolsHeader)

	slog.Default().InfoContext(ctx, "custom-message trigger",
		slog.String("triggerSource", triggerSource),
		slog.String("userName", userName),
	)

	var emailSubject, emailMessage, smsMessage string

	switch triggerSource {
	case "CustomMessage_SignUp":
		emailSubject = "Mabuhay! Welcome to TowCommand - Verify Your Account"
		emailMessage = "Mabuhay! Welcome to TowCommand. Your verification code is {####}. " +
			"Please enter this code to verify your account."
		smsMessage = "Mabuhay! TowCommand verification code: {####}"

	case "CustomMessage_ForgotPassword":
		emailSubject = "TowCommand - Password Reset Code"
		emailMessage = "Your TowCommand password reset code is {####}. " +
			"If you did not request this, please ignore this message."
		smsMessage = "TowCommand password reset code: {####}"

	case "CustomMessage_ResendCode":
		emailSubject = "TowCommand - Verification Code"
		emailMessage = "Your TowCommand verification code is {####}. " +
			"Please enter this code to verify your account."
		smsMessage = "TowCommand verification code: {####}"

	case "CustomMessage_AdminCreateUser":
		emailSubject = "Mabuhay! Welcome to TowCommand"
		emailMessage = "Mabuhay! Your TowCommand account has been created. " +
			"Your temporary password is {####}. Please log in and change it."
		smsMessage = "Mabuhay! TowCommand temporary password: {####}"

	default:
		slog.Default().InfoContext(ctx, "no custom template for trigger source, using defaults",
			slog.String("triggerSource", triggerSource),
		)
		return *event, nil
	}

	return handler.CustomizeMessage(event, emailSubject, emailMessage, smsMessage), nil
}
